{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "96a0be15-746a-4b0c-9daa-e99972b17c67",
   "metadata": {},
   "source": [
    "# SAIGE Prototype\n",
    "\n",
    "This notebook demonstrates a prototype of [SAIGE](https://saigegit.github.io/SAIGE-doc/) that supports reading from VCF Zarr stores.\n",
    "\n",
    "Run the `setup.sh` script before using this notebook to install SAIGE and create a Conda environment for SAIGE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "96de7c24-13b6-4566-9ece-9011cbc27ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting tstrait\n",
      "  Using cached tstrait-0.1.0-py3-none-any.whl.metadata (2.3 kB)\n",
      "Collecting numpy>=1.20.3 (from tstrait)\n",
      "  Downloading numpy-2.2.1-cp310-cp310-macosx_14_0_arm64.whl.metadata (62 kB)\n",
      "Collecting numba>=0.57.0 (from tstrait)\n",
      "  Using cached numba-0.60.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (2.7 kB)\n",
      "Collecting pandas>=1.0 (from tstrait)\n",
      "  Using cached pandas-2.2.3-cp310-cp310-macosx_11_0_arm64.whl.metadata (89 kB)\n",
      "Collecting tskit>=0.5.5 (from tstrait)\n",
      "  Downloading tskit-0.6.0-cp310-cp310-macosx_10_9_universal2.whl.metadata (2.0 kB)\n",
      "Collecting llvmlite<0.44,>=0.43.0dev0 (from numba>=0.57.0->tstrait)\n",
      "  Using cached llvmlite-0.43.0-cp310-cp310-macosx_11_0_arm64.whl.metadata (4.8 kB)\n",
      "Collecting numpy>=1.20.3 (from tstrait)\n",
      "  Downloading numpy-2.0.2-cp310-cp310-macosx_14_0_arm64.whl.metadata (60 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /opt/miniconda3/envs/vcf-zarr-publication/lib/python3.10/site-packages (from pandas>=1.0->tstrait) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/miniconda3/envs/vcf-zarr-publication/lib/python3.10/site-packages (from pandas>=1.0->tstrait) (2024.2)\n",
      "Collecting tzdata>=2022.7 (from pandas>=1.0->tstrait)\n",
      "  Using cached tzdata-2024.2-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "Requirement already satisfied: jsonschema>=3.0.0 in /opt/miniconda3/envs/vcf-zarr-publication/lib/python3.10/site-packages (from tskit>=0.5.5->tstrait) (4.23.0)\n",
      "Collecting svgwrite>=1.1.10 (from tskit>=0.5.5->tstrait)\n",
      "  Using cached svgwrite-1.4.3-py3-none-any.whl.metadata (8.8 kB)\n",
      "Requirement already satisfied: attrs>=22.2.0 in /opt/miniconda3/envs/vcf-zarr-publication/lib/python3.10/site-packages (from jsonschema>=3.0.0->tskit>=0.5.5->tstrait) (24.3.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /opt/miniconda3/envs/vcf-zarr-publication/lib/python3.10/site-packages (from jsonschema>=3.0.0->tskit>=0.5.5->tstrait) (2024.10.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in /opt/miniconda3/envs/vcf-zarr-publication/lib/python3.10/site-packages (from jsonschema>=3.0.0->tskit>=0.5.5->tstrait) (0.35.1)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in /opt/miniconda3/envs/vcf-zarr-publication/lib/python3.10/site-packages (from jsonschema>=3.0.0->tskit>=0.5.5->tstrait) (0.22.3)\n",
      "Requirement already satisfied: six>=1.5 in /opt/miniconda3/envs/vcf-zarr-publication/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas>=1.0->tstrait) (1.17.0)\n",
      "Using cached tstrait-0.1.0-py3-none-any.whl (16 kB)\n",
      "Using cached numba-0.60.0-cp310-cp310-macosx_11_0_arm64.whl (2.7 MB)\n",
      "Downloading numpy-2.0.2-cp310-cp310-macosx_14_0_arm64.whl (5.3 MB)\n",
      "\u001b[2K   \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.3/5.3 MB\u001b[0m \u001b[31m11.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m31m15.9 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hUsing cached pandas-2.2.3-cp310-cp310-macosx_11_0_arm64.whl (11.3 MB)\n",
      "Downloading tskit-0.6.0-cp310-cp310-macosx_10_9_universal2.whl (737 kB)\n",
      "\u001b[2K   \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m737.9/737.9 kB\u001b[0m \u001b[31m11.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hUsing cached llvmlite-0.43.0-cp310-cp310-macosx_11_0_arm64.whl (28.8 MB)\n",
      "Using cached svgwrite-1.4.3-py3-none-any.whl (67 kB)\n",
      "Using cached tzdata-2024.2-py2.py3-none-any.whl (346 kB)\n",
      "Installing collected packages: tzdata, svgwrite, numpy, llvmlite, pandas, numba, tskit, tstrait\n",
      "Successfully installed llvmlite-0.43.0 numba-0.60.0 numpy-2.0.2 pandas-2.2.3 svgwrite-1.4.3 tskit-0.6.0 tstrait-0.1.0 tzdata-2024.2\n"
     ]
    }
   ],
   "source": [
    "!pip install tstrait"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8c9ba97-1b2a-4cfd-ab41-e372205a84d5",
   "metadata": {},
   "source": [
    "### Simulate phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "082226e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tskit\n",
    "import tstrait\n",
    "\n",
    "ts = tskit.load('../scaling/data/chr21_10_5.ts')\n",
    "model = tstrait.trait_model(distribution='normal', mean=0, var=1)\n",
    "sim_result = tstrait.sim_phenotype(ts=ts, model=model, h2=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "62639928-992d-4c63-8cb4-ad08d1d2747f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>site_id</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>causal_allele</th>\n",
       "      <th>allele_freq</th>\n",
       "      <th>trait_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23605820</td>\n",
       "      <td>654107</td>\n",
       "      <td>0.372912</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   position  site_id  effect_size causal_allele  allele_freq  trait_id\n",
       "0  23605820   654107     0.372912             A     0.000035         0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_result.trait"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0b0533f2-1228-4b12-8dca-e8589c168557",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait_id</th>\n",
       "      <th>individual_id</th>\n",
       "      <th>genetic_value</th>\n",
       "      <th>environmental_noise</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.006036</td>\n",
       "      <td>-0.006036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.000652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001182</td>\n",
       "      <td>0.001182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.001402</td>\n",
       "      <td>-0.001402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.001009</td>\n",
       "      <td>-0.001009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286713</th>\n",
       "      <td>0</td>\n",
       "      <td>286713</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000769</td>\n",
       "      <td>0.000769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286714</th>\n",
       "      <td>0</td>\n",
       "      <td>286714</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000580</td>\n",
       "      <td>-0.000580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286715</th>\n",
       "      <td>0</td>\n",
       "      <td>286715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005226</td>\n",
       "      <td>0.005226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286716</th>\n",
       "      <td>0</td>\n",
       "      <td>286716</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.003341</td>\n",
       "      <td>-0.003341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286717</th>\n",
       "      <td>0</td>\n",
       "      <td>286717</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000051</td>\n",
       "      <td>-0.000051</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286718 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        trait_id  individual_id  genetic_value  environmental_noise  phenotype\n",
       "0              0              0            0.0            -0.006036  -0.006036\n",
       "1              0              1            0.0             0.000652   0.000652\n",
       "2              0              2            0.0             0.001182   0.001182\n",
       "3              0              3            0.0            -0.001402  -0.001402\n",
       "4              0              4            0.0            -0.001009  -0.001009\n",
       "...          ...            ...            ...                  ...        ...\n",
       "286713         0         286713            0.0             0.000769   0.000769\n",
       "286714         0         286714            0.0            -0.000580  -0.000580\n",
       "286715         0         286715            0.0             0.005226   0.005226\n",
       "286716         0         286716            0.0            -0.003341  -0.003341\n",
       "286717         0         286717            0.0            -0.000051  -0.000051\n",
       "\n",
       "[286718 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sim_result.phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "76c37244-f4ba-42cf-b6df-8251344cae2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait_id</th>\n",
       "      <th>individual_id</th>\n",
       "      <th>genetic_value</th>\n",
       "      <th>environmental_noise</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.006036</td>\n",
       "      <td>-0.006036</td>\n",
       "      <td>tsk_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>tsk_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001182</td>\n",
       "      <td>0.001182</td>\n",
       "      <td>tsk_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.001402</td>\n",
       "      <td>-0.001402</td>\n",
       "      <td>tsk_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.001009</td>\n",
       "      <td>-0.001009</td>\n",
       "      <td>tsk_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286713</th>\n",
       "      <td>0</td>\n",
       "      <td>286713</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000769</td>\n",
       "      <td>0.000769</td>\n",
       "      <td>tsk_286713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286714</th>\n",
       "      <td>0</td>\n",
       "      <td>286714</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000580</td>\n",
       "      <td>-0.000580</td>\n",
       "      <td>tsk_286714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286715</th>\n",
       "      <td>0</td>\n",
       "      <td>286715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005226</td>\n",
       "      <td>0.005226</td>\n",
       "      <td>tsk_286715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286716</th>\n",
       "      <td>0</td>\n",
       "      <td>286716</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.003341</td>\n",
       "      <td>-0.003341</td>\n",
       "      <td>tsk_286716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286717</th>\n",
       "      <td>0</td>\n",
       "      <td>286717</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000051</td>\n",
       "      <td>-0.000051</td>\n",
       "      <td>tsk_286717</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286718 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        trait_id  individual_id  genetic_value  environmental_noise  \\\n",
       "0              0              0            0.0            -0.006036   \n",
       "1              0              1            0.0             0.000652   \n",
       "2              0              2            0.0             0.001182   \n",
       "3              0              3            0.0            -0.001402   \n",
       "4              0              4            0.0            -0.001009   \n",
       "...          ...            ...            ...                  ...   \n",
       "286713         0         286713            0.0             0.000769   \n",
       "286714         0         286714            0.0            -0.000580   \n",
       "286715         0         286715            0.0             0.005226   \n",
       "286716         0         286716            0.0            -0.003341   \n",
       "286717         0         286717            0.0            -0.000051   \n",
       "\n",
       "        phenotype   sample_id  \n",
       "0       -0.006036       tsk_0  \n",
       "1        0.000652       tsk_1  \n",
       "2        0.001182       tsk_2  \n",
       "3       -0.001402       tsk_3  \n",
       "4       -0.001009       tsk_4  \n",
       "...           ...         ...  \n",
       "286713   0.000769  tsk_286713  \n",
       "286714  -0.000580  tsk_286714  \n",
       "286715   0.005226  tsk_286715  \n",
       "286716  -0.003341  tsk_286716  \n",
       "286717  -0.000051  tsk_286717  \n",
       "\n",
       "[286718 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype = sim_result.phenotype\n",
    "phenotype['sample_id'] = 'tsk_' + phenotype['individual_id'].astype(str)\n",
    "phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d576714c-50ae-46fc-82a9-964768a123b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save phenotype data to disk in the format that SAIGE expects.\n",
    "phenotype[['sample_id', 'phenotype']].to_csv('chr21_10_5.phenotypes.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d35e5e64-af12-4126-a500-7e0b0a5b11d1",
   "metadata": {},
   "source": [
    "### SAIGE workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "41603ff5-488b-4933-b909-ced023276321",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a5.12 M1 (25 Jun 2024)              www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./chr21_10_4.log.\n",
      "Options in effect:\n",
      "  --make-bed\n",
      "  --max-alleles 2\n",
      "  --out ./chr21_10_4\n",
      "  --vcf ../scaling/data/chr21_10_4.vcf.gz\n",
      "\n",
      "Start time: Sat Jan 11 20:12:48 2025\n",
      "8192 MiB RAM detected; reserving 4096 MiB for main workspace.\n",
      "Using up to 8 compute threads.\n",
      "--vcf: 863998 variants scanned.\n",
      "--vcf: ./chr21_10_4-temporary.pgen + ./chr21_10_4-temporary.pvar.zst +\n",
      "./chr21_10_4-temporary.psam written.\n",
      "10000 samples (0 females, 0 males, 10000 ambiguous; 10000 founders) loaded from\n",
      "./chr21_10_4-temporary.psam.\n",
      "856315 out of 863998 variants loaded from ./chr21_10_4-temporary.pvar.zst.\n",
      "Note: No phenotype data present.\n",
      "856315 variants remaining after main filters.\n",
      "Writing ./chr21_10_4.fam ... done.\n",
      "Writing ./chr21_10_4.bim ... done.\n",
      "done.hr21_10_4.bed ... 0%\n",
      "End time: Sat Jan 11 20:13:26 2025\n"
     ]
    }
   ],
   "source": [
    "!plink2 --vcf ../scaling/data/chr21_10_4.vcf.gz --make-bed --out ./chr21_10_4 --max-alleles 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b55b1c66-0e94-4edd-87ba-aa4f3509cdae",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: optparse\n",
      "Warning message:\n",
      "package ‘optparse’ was built under R version 4.3.3 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R version 4.3.0 (2023-04-21)\n",
      "Platform: x86_64-apple-darwin13.4.0 (64-bit)\n",
      "Running under: macOS Big Sur ... 10.16\n",
      "\n",
      "Matrix products: default\n",
      "BLAS:   /opt/miniconda3/envs/RSAIGE/lib/libblis.4.0.0.dylib \n",
      "LAPACK: /opt/miniconda3/envs/RSAIGE/lib/liblapack.3.9.0.dylib\n",
      "\n",
      "locale:\n",
      "[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8\n",
      "\n",
      "time zone: UTC\n",
      "tzcode source: system (macOS)\n",
      "\n",
      "attached base packages:\n",
      "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
      "\n",
      "other attached packages:\n",
      "[1] optparse_1.7.5 SAIGE_1.3.6   \n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      "[1] compiler_4.3.0     Matrix_1.6-5       Rcpp_1.0.13        getopt_1.20.4     \n",
      "[5] grid_4.3.0         data.table_1.15.2  RcppParallel_5.1.9 lattice_0.22-6    \n",
      "$plinkFile\n",
      "[1] \"./chr21_10_4\"\n",
      "\n",
      "$bedFile\n",
      "[1] \"\"\n",
      "\n",
      "$bimFile\n",
      "[1] \"\"\n",
      "\n",
      "$famFile\n",
      "[1] \"\"\n",
      "\n",
      "$phenoFile\n",
      "[1] \"./chr21_10_5.phenotypes.txt\"\n",
      "\n",
      "$phenoCol\n",
      "[1] \"phenotype\"\n",
      "\n",
      "$traitType\n",
      "[1] \"quantitative\"\n",
      "\n",
      "$invNormalize\n",
      "[1] TRUE\n",
      "\n",
      "$covarColList\n",
      "[1] \"\"\n",
      "\n",
      "$qCovarColList\n",
      "[1] \"\"\n",
      "\n",
      "$sampleIDColinphenoFile\n",
      "[1] \"sample_id\"\n",
      "\n",
      "$tol\n",
      "[1] 0.02\n",
      "\n",
      "$maxiter\n",
      "[1] 20\n",
      "\n",
      "$tolPCG\n",
      "[1] 1e-05\n",
      "\n",
      "$maxiterPCG\n",
      "[1] 500\n",
      "\n",
      "$nThreads\n",
      "[1] 24\n",
      "\n",
      "$SPAcutoff\n",
      "[1] 2\n",
      "\n",
      "$numRandomMarkerforVarianceRatio\n",
      "[1] 30\n",
      "\n",
      "$skipModelFitting\n",
      "[1] FALSE\n",
      "\n",
      "$skipVarianceRatioEstimation\n",
      "[1] FALSE\n",
      "\n",
      "$memoryChunk\n",
      "[1] 2\n",
      "\n",
      "$tauInit\n",
      "[1] \"0,0\"\n",
      "\n",
      "$LOCO\n",
      "[1] TRUE\n",
      "\n",
      "$isLowMemLOCO\n",
      "[1] FALSE\n",
      "\n",
      "$traceCVcutoff\n",
      "[1] 0.0025\n",
      "\n",
      "$nrun\n",
      "[1] 30\n",
      "\n",
      "$ratioCVcutoff\n",
      "[1] 0.001\n",
      "\n",
      "$outputPrefix\n",
      "[1] \"./chr21_10_4.model\"\n",
      "\n",
      "$outputPrefix_varRatio\n",
      "[1] \"\"\n",
      "\n",
      "$IsOverwriteVarianceRatioFile\n",
      "[1] TRUE\n",
      "\n",
      "$sparseGRMFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMSampleIDFile\n",
      "[1] \"\"\n",
      "\n",
      "$isCateVarianceRatio\n",
      "[1] FALSE\n",
      "\n",
      "$relatednessCutoff\n",
      "[1] 0\n",
      "\n",
      "$cateVarRatioMinMACVecExclude\n",
      "[1] \"10,20.5\"\n",
      "\n",
      "$cateVarRatioMaxMACVecInclude\n",
      "[1] \"20.5\"\n",
      "\n",
      "$isCovariateTransform\n",
      "[1] TRUE\n",
      "\n",
      "$isDiagofKinSetAsOne\n",
      "[1] FALSE\n",
      "\n",
      "$useSparseGRMtoFitNULL\n",
      "[1] FALSE\n",
      "\n",
      "$useSparseGRMforVarRatio\n",
      "[1] FALSE\n",
      "\n",
      "$minMAFforGRM\n",
      "[1] 0.01\n",
      "\n",
      "$maxMissingRateforGRM\n",
      "[1] 0.15\n",
      "\n",
      "$minCovariateCount\n",
      "[1] -1\n",
      "\n",
      "$includeNonautoMarkersforVarRatio\n",
      "[1] FALSE\n",
      "\n",
      "$FemaleOnly\n",
      "[1] FALSE\n",
      "\n",
      "$MaleOnly\n",
      "[1] FALSE\n",
      "\n",
      "$sexCol\n",
      "[1] \"\"\n",
      "\n",
      "$FemaleCode\n",
      "[1] \"1\"\n",
      "\n",
      "$MaleCode\n",
      "[1] \"0\"\n",
      "\n",
      "$isCovariateOffset\n",
      "[1] TRUE\n",
      "\n",
      "$SampleIDIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$help\n",
      "[1] FALSE\n",
      "\n",
      "tauInit  is  0 0 \n",
      "cateVarRatioMinMACVecExclude  is  10 20.5 \n",
      "cateVarRatioMaxMACVecInclude  is  20.5 \n",
      "The variance ratio file  ./chr21_10_4.model.varianceRatio.txt  already exists. IsOverwriteVarianceRatioFile=TRUE so the file will be overwritten\n",
      "24  threads will be used  \n",
      "WARNING: Genetic variants needs to be ordered by chromosome and position in the Plink file\n",
      "chromosomeStartIndexVec:  0 NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA \n",
      "chromosomeEndIndexVec:  856314 NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA \n",
      "WARNING: The number of autosomal chromosomes is less than 2 and leave-one-chromosome-out can't be conducted! \n",
      "10000  samples have genotypes\n",
      "formula is  phenotype~ 1 \n",
      "286718  samples have non-missing phenotypes\n",
      "10000  samples will be used for analysis\n",
      "Perform the inverse nomalization for  phenotype \n",
      "       IID X.Intercept.  phenotype IndexPheno IndexGeno\n",
      "1    tsk_0            1 -1.6338041          1         1\n",
      "2    tsk_1            1  0.1632774          2         2\n",
      "1113 tsk_2            1  0.3140277          3         3\n",
      "2224 tsk_3            1 -0.3968203          4         4\n",
      "3335 tsk_4            1 -0.2951229          5         5\n",
      "4446 tsk_5            1 -0.8590734          6         6\n",
      "[1] \"No covariate is includes so isCovariateOffset = FALSE\"\n",
      "geno.g_minMACVarRatio 20\n",
      "geno.g_maxMACVarRatio -1\n",
      "Markers in the Plink file with MAF <  0.01  will be removed before constructing GRM\n",
      "Markers in the Plink file with missing rate >  0.15  will be removed before constructing GRM\n",
      "phenotype  is a quantitative trait\n",
      "glm:\n",
      "\n",
      "Call:  glm(formula = formula.new, family = gaussian(link = \"identity\"), \n",
      "    data = data.new)\n",
      "\n",
      "Coefficients:\n",
      "(Intercept)  \n",
      " -8.882e-17  \n",
      "\n",
      "Degrees of Freedom: 9999 Total (i.e. Null);  9999 Residual\n",
      "Null Deviance:\t    9999 \n",
      "Residual Deviance: 9999 \tAIC: 28380\n",
      "   user  system elapsed \n",
      "  3.196   0.193   2.291 \n",
      "Start fitting the NULL GLMM\n",
      "   user  system elapsed \n",
      "  3.254   0.193   2.350 \n",
      "[1] \"Start reading genotype plink file here\"\n",
      "nbyte: 2500\n",
      "nbyte: 2500\n",
      "reserve: 2142500096\n",
      "\n",
      "M: 856315, N: 10000\n",
      "setgeno mark1\n",
      "setgeno mark2\n",
      "125522 markers with MAF >= 0.01 and missing rate <= 0.15\n",
      "time: 34105\n",
      "[1] \"Genotype reading is done\"\n",
      "[1] \"Genotype reading is done\"\n",
      "Fixed-effect coefficients:\n",
      "  (Intercept) \n",
      "-8.881784e-17 \n",
      "initial tau is  1 0 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Tau:\n",
      "[1] 1 0\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] 2.861023e-09\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Variance component estimates:\n",
      "[1] 0.9999677 0.0000000\n",
      "\n",
      "Iteration  1 :\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Tau:\n",
      "[1] 0.9999677 0.0000000\n",
      "Fixed-effect coefficients:\n",
      "              [,1]\n",
      "[1,] -5.531104e-09\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "cov:  9.999625e-05 \n",
      "Variance component estimates:\n",
      "[1] 0.9984896 0.0000000\n",
      "Fixed-effect coefficients:\n",
      "              [,1]\n",
      "[1,] -5.531104e-09\n",
      "\n",
      "Final  0.9984896 0 :\n",
      "t_end_null - t_begin, fitting the NULL model without LOCO took\n",
      "   user  system elapsed \n",
      "178.098   3.200  55.091 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "Tau:\n",
      "[1] 0.9984896 0.0000000\n",
      "Fixed-effect coefficients:\n",
      "             [,1]\n",
      "[1,] 1.142682e-09\n",
      "   user  system elapsed \n",
      "181.465   3.466  57.631 \n",
      "t_end - t_begin, fitting the NULL model took\n",
      "   user  system elapsed \n",
      "178.269   3.273  55.340 \n",
      "Start estimating variance ratios\n",
      "\n",
      "Family: gaussian \n",
      "Link function: identity \n",
      "\n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "length(MACvector):  125522 \n",
      "length(MACvector_forVarRatio):  293 \n",
      "Only one variance ratio will be estimated using randomly selected markers with MAC >= 20\n",
      "47 th marker in geno  1 \n",
      "MAC:  1302 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "206 th marker in geno  1 \n",
      "MAC:  1756 \n",
      "G0 0 0 0 0 1 0 1 0 1 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "169 th marker in geno  1 \n",
      "MAC:  106 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "152 th marker in geno  1 \n",
      "MAC:  232 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "134 th marker in geno  1 \n",
      "MAC:  8014 \n",
      "G0 0 1 2 0 1 1 1 0 2 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "174 th marker in geno  1 \n",
      "MAC:  8712 \n",
      "G0 0 1 0 0 2 1 1 2 2 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "92 th marker in geno  1 \n",
      "MAC:  95 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "42 th marker in geno  1 \n",
      "MAC:  48 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "228 th marker in geno  1 \n",
      "MAC:  80 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "238 th marker in geno  1 \n",
      "MAC:  24 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "6 th marker in geno  1 \n",
      "MAC:  843 \n",
      "G0 0 0 0 0 1 1 0 0 1 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "240 th marker in geno  1 \n",
      "MAC:  2087 \n",
      "G0 0 1 2 1 1 2 2 2 2 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "291 th marker in geno  1 \n",
      "MAC:  964 \n",
      "G0 0 0 0 0 0 0 0 1 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "213 th marker in geno  1 \n",
      "MAC:  86 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "201 th marker in geno  1 \n",
      "MAC:  424 \n",
      "G0 0 0 1 0 0 0 1 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "186 th marker in geno  1 \n",
      "MAC:  1023 \n",
      "G0 0 0 0 0 0 0 0 0 1 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "89 th marker in geno  1 \n",
      "MAC:  67 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "72 th marker in geno  1 \n",
      "MAC:  21 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "232 th marker in geno  1 \n",
      "MAC:  121 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "50 th marker in geno  1 \n",
      "MAC:  34 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "27 th marker in geno  1 \n",
      "MAC:  3014 \n",
      "G0 0 1 0 0 1 0 0 0 0 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "9 th marker in geno  1 \n",
      "MAC:  29 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "131 th marker in geno  1 \n",
      "MAC:  31 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "167 th marker in geno  1 \n",
      "MAC:  36 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "109 th marker in geno  1 \n",
      "MAC:  802 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "106 th marker in geno  1 \n",
      "MAC:  1819 \n",
      "G0 0 0 0 0 0 0 0 1 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "212 th marker in geno  1 \n",
      "MAC:  23 \n",
      "G0 0 0 0 0 0 0 0 1 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "41 th marker in geno  1 \n",
      "MAC:  5640 \n",
      "G0 0 0 1 0 0 1 1 0 0 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "64 th marker in geno  1 \n",
      "MAC:  67 \n",
      "G0 0 0 0 0 0 0 0 0 0 0 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "82 th marker in geno  1 \n",
      "MAC:  2630 \n",
      "G0 0 0 1 0 0 1 1 0 0 1 \n",
      "CHR  1 \n",
      "iter from getPCG1ofSigmaAndVector 1\n",
      "CV for variance ratio estimate using  30  markers is  1.921268e-09  <  0.001 \n",
      "varRatio_null 1.001513 \n",
      "        V1   V2 V3\n",
      "1 1.001513 null  1\n",
      "closed the plinkFile!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "export PATH=\"/opt/miniconda3/bin:$PATH\"\n",
    "conda run -n RSAIGE Rscript SAIGE/extdata/step1_fitNULLGLMM.R     \\\n",
    "        --plinkFile=./chr21_10_4  \\\n",
    "        --useSparseGRMtoFitNULL=FALSE    \\\n",
    "        --phenoFile=./chr21_10_5.phenotypes.txt \\\n",
    "        --phenoCol=phenotype \\\n",
    "        --sampleIDColinphenoFile=sample_id \\\n",
    "        --invNormalize=TRUE     \\\n",
    "        --traitType=quantitative        \\\n",
    "        --outputPrefix=./chr21_10_4.model \\\n",
    "        --nThreads=24   \\\n",
    "        --IsOverwriteVarianceRatioFile=TRUE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "395cecfd-a6e5-454d-8439-b89717748bdf",
   "metadata": {},
   "source": [
    "Here we run [step 2](https://saigegit.github.io/SAIGE-doc/docs/single_step2.html) of the single-variant association test using the VCF data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2d96ede6-859d-429b-b6f2-4838e1f52fc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R version 4.3.0 (2023-04-21)\n",
      "Platform: x86_64-apple-darwin13.4.0 (64-bit)\n",
      "Running under: macOS Big Sur ... 10.16\n",
      "\n",
      "Matrix products: default\n",
      "BLAS:   /opt/miniconda3/envs/RSAIGE/lib/libblis.4.0.0.dylib \n",
      "LAPACK: /opt/miniconda3/envs/RSAIGE/lib/liblapack.3.9.0.dylib\n",
      "\n",
      "locale:\n",
      "[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8\n",
      "\n",
      "time zone: UTC\n",
      "tzcode source: system (macOS)\n",
      "\n",
      "attached base packages:\n",
      "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
      "\n",
      "other attached packages:\n",
      "[1] data.table_1.15.2   optparse_1.7.5      RhpcBLASctl_0.23-42\n",
      "[4] SAIGE_1.3.6        \n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      "[1] compiler_4.3.0     Matrix_1.6-5       Rcpp_1.0.13        getopt_1.20.4     \n",
      "[5] grid_4.3.0         RcppParallel_5.1.9 lattice_0.22-6    \n",
      "$vcfFile\n",
      "[1] \"../scaling/data/chr21_10_4.bcf\"\n",
      "\n",
      "$vcfFileIndex\n",
      "[1] \"../scaling/data/chr21_10_4.bcf.csi\"\n",
      "\n",
      "$vcfField\n",
      "[1] \"GT\"\n",
      "\n",
      "$vczFile\n",
      "[1] \"\"\n",
      "\n",
      "$savFile\n",
      "[1] \"\"\n",
      "\n",
      "$savFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$bgenFile\n",
      "[1] \"\"\n",
      "\n",
      "$bgenFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$sampleFile\n",
      "[1] \"\"\n",
      "\n",
      "$bedFile\n",
      "[1] \"\"\n",
      "\n",
      "$bimFile\n",
      "[1] \"\"\n",
      "\n",
      "$famFile\n",
      "[1] \"\"\n",
      "\n",
      "$AlleleOrder\n",
      "[1] \"alt-first\"\n",
      "\n",
      "$idstoIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$rangestoIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$chrom\n",
      "[1] \"1\"\n",
      "\n",
      "$is_imputed_data\n",
      "[1] FALSE\n",
      "\n",
      "$minMAF\n",
      "[1] 0\n",
      "\n",
      "$minMAC\n",
      "[1] 20\n",
      "\n",
      "$minGroupMAC_in_BurdenTest\n",
      "[1] 5\n",
      "\n",
      "$minInfo\n",
      "[1] 0\n",
      "\n",
      "$maxMissing\n",
      "[1] 0.15\n",
      "\n",
      "$impute_method\n",
      "[1] \"best_guess\"\n",
      "\n",
      "$LOCO\n",
      "[1] FALSE\n",
      "\n",
      "$GMMATmodelFile\n",
      "[1] \"./chr21_10_4.model.rda\"\n",
      "\n",
      "$varianceRatioFile\n",
      "[1] \"./chr21_10_4.model.varianceRatio.txt\"\n",
      "\n",
      "$SAIGEOutputFile\n",
      "[1] \"./chr21_10_4.bcf_results.txt\"\n",
      "\n",
      "$markers_per_chunk\n",
      "[1] 10000\n",
      "\n",
      "$groups_per_chunk\n",
      "[1] 100\n",
      "\n",
      "$is_output_moreDetails\n",
      "[1] TRUE\n",
      "\n",
      "$is_overwrite_output\n",
      "[1] TRUE\n",
      "\n",
      "$maxMAF_in_groupTest\n",
      "[1] \"0.0001,0.001,0.01\"\n",
      "\n",
      "$maxMAC_in_groupTest\n",
      "[1] \"0\"\n",
      "\n",
      "$annotation_in_groupTest\n",
      "[1] \"lof,missense;lof,missense;lof;synonymous\"\n",
      "\n",
      "$groupFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMSampleIDFile\n",
      "[1] \"\"\n",
      "\n",
      "$relatednessCutoff\n",
      "[1] 0\n",
      "\n",
      "$sampleFile_male\n",
      "[1] \"\"\n",
      "\n",
      "$X_PARregion\n",
      "[1] \"\"\n",
      "\n",
      "$is_rewrite_XnonPAR_forMales\n",
      "[1] FALSE\n",
      "\n",
      "$MACCutoff_to_CollapseUltraRare\n",
      "[1] 10\n",
      "\n",
      "$cateVarRatioMinMACVecExclude\n",
      "[1] \"10,20.5\"\n",
      "\n",
      "$cateVarRatioMaxMACVecInclude\n",
      "[1] \"20.5\"\n",
      "\n",
      "$weights.beta\n",
      "[1] \"1,25\"\n",
      "\n",
      "$r.corr\n",
      "[1] 0\n",
      "\n",
      "$markers_per_chunk_in_groupTest\n",
      "[1] 100\n",
      "\n",
      "$condition\n",
      "[1] \"\"\n",
      "\n",
      "$SPAcutoff\n",
      "[1] 2\n",
      "\n",
      "$dosage_zerod_cutoff\n",
      "[1] 0.2\n",
      "\n",
      "$dosage_zerod_MAC_cutoff\n",
      "[1] 10\n",
      "\n",
      "$is_single_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_no_weight_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_output_markerList_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_Firth_beta\n",
      "[1] TRUE\n",
      "\n",
      "$pCutoffforFirth\n",
      "[1] 0.05\n",
      "\n",
      "$is_fastTest\n",
      "[1] FALSE\n",
      "\n",
      "$max_MAC_for_ER\n",
      "[1] 4\n",
      "\n",
      "$subSampleFile\n",
      "[1] \"\"\n",
      "\n",
      "$help\n",
      "[1] FALSE\n",
      "\n",
      "[1] \"opt$r.corr\"\n",
      "[1] 0\n",
      "dosage_zerod_cutoff  0.2 \n",
      "Any dosages <=  0.2  for genetic variants with MAC <=  10  are set to be 0.\n",
      "single-variant association test will be performed\n",
      "[1] \"Leave-one-chromosome-out is not applied\"\n",
      "[1] \"LOCO = FASLE and leave-one-chromosome-out is not applied\"\n",
      "variance Ratio null is  1.001513 \n",
      "Please note the argument vcfFileIndex will not be used in future versions because the vcf index file must has the name 'vcfFile'.csi\n",
      "dosageFile type is  vcf \n",
      "Open VCF done\n",
      "To read the field GT\n",
      "Number of meta lines in the vcf file (lines starting with ##): 7\n",
      "Number of samples in the vcf file: 10000\n",
      "Setting position of samples in VCF files....\n",
      "m_N 10000\n",
      "(2025-01-12 00:11:35.036574) ---- Analyzing Chunk 1 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3333 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  2.675   0.222   3.046 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:36.07) ---- Analyzing Chunk 2 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3421 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  3.669   0.228   4.046 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:37.067852) ---- Analyzing Chunk 3 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3182 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  4.632   0.234   5.015 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:38.03582) ---- Analyzing Chunk 4 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3268 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  5.581   0.240   5.971 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:38.992161) ---- Analyzing Chunk 5 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3040 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  6.507   0.246   6.904 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:39.923285) ---- Analyzing Chunk 6 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3117 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  7.446   0.252   7.849 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:40.869129) ---- Analyzing Chunk 7 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3051 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  8.409   0.258   8.818 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:41.840281) ---- Analyzing Chunk 8 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3044 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  9.387   0.264   9.801 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:42.822495) ---- Analyzing Chunk 9 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3038 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 10.325   0.270  10.745 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:43.769986) ---- Analyzing Chunk 10 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3237 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 11.302   0.276  11.728 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:44.748836) ---- Analyzing Chunk 11 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3119 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 12.241   0.281  12.673 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:45.694731) ---- Analyzing Chunk 12 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3276 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 13.198   0.287  13.635 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:46.654594) ---- Analyzing Chunk 13 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3181 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 14.144   0.292  14.588 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:47.608978) ---- Analyzing Chunk 14 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3029 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 15.087   0.300  15.540 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:48.563796) ---- Analyzing Chunk 15 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2879 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 16.002   0.308  16.463 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:49.483059) ---- Analyzing Chunk 16 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3015 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 16.935   0.315  17.403 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:50.42398) ---- Analyzing Chunk 17 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3072 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 17.899   0.322  18.374 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:51.393893) ---- Analyzing Chunk 18 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3157 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 18.879   0.328  19.359 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:52.380294) ---- Analyzing Chunk 19 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3475 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 19.904   0.336  20.393 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:53.413688) ---- Analyzing Chunk 20 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3226 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 20.866   0.343  21.362 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:54.385782) ---- Analyzing Chunk 21 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3284 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 21.862   0.351  22.366 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:55.387072) ---- Analyzing Chunk 22 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3073 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 22.796   0.357  23.307 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:56.32774) ---- Analyzing Chunk 23 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2956 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 23.723   0.363  24.240 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:57.262277) ---- Analyzing Chunk 24 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3108 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 24.710   0.374  25.240 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:58.260603) ---- Analyzing Chunk 25 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3073 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 25.658   0.380  26.195 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:11:59.214352) ---- Analyzing Chunk 26 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2642 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 26.580   0.386  27.123 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:00.141962) ---- Analyzing Chunk 27 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2815 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 27.510   0.393  28.065 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:01.08408) ---- Analyzing Chunk 28 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2967 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 28.452   0.398  29.012 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:02.031574) ---- Analyzing Chunk 29 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2936 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 29.362   0.403  29.928 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:02.953336) ---- Analyzing Chunk 30 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2634 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 30.275   0.408  30.847 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:03.867779) ---- Analyzing Chunk 31 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3025 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 31.208   0.414  31.786 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:04.809427) ---- Analyzing Chunk 32 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3067 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 32.145   0.419  32.729 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:05.748487) ---- Analyzing Chunk 33 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3308 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 33.118   0.424  33.707 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:06.726284) ---- Analyzing Chunk 34 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3047 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 34.055   0.430  34.649 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:07.668474) ---- Analyzing Chunk 35 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3109 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 35.007   0.436  35.608 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:08.627167) ---- Analyzing Chunk 36 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2790 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 35.933   0.441  36.539 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:09.558452) ---- Analyzing Chunk 37 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3031 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 36.881   0.447  37.494 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:10.513539) ---- Analyzing Chunk 38 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3079 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 37.823   0.452  38.442 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:11.46116) ---- Analyzing Chunk 39 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3113 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 38.774   0.457  39.398 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:12.418937) ---- Analyzing Chunk 40 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3109 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 39.749   0.463  40.379 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:13.399031) ---- Analyzing Chunk 41 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3324 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 40.712   0.470  41.352 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:14.372549) ---- Analyzing Chunk 42 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3662 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 41.830   0.478  42.478 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:15.497583) ---- Analyzing Chunk 43 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2965 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 42.784   0.483  43.437 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:16.457056) ---- Analyzing Chunk 44 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2883 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 43.699   0.488  44.358 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:17.376947) ---- Analyzing Chunk 45 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3064 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 44.659   0.493  45.323 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:18.342155) ---- Analyzing Chunk 46 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3094 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 45.603   0.499  46.272 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:19.292097) ---- Analyzing Chunk 47 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3085 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 46.559   0.504  47.234 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:20.25328) ---- Analyzing Chunk 48 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2926 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 47.516   0.509  48.195 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:21.215144) ---- Analyzing Chunk 49 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2852 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 48.446   0.514  49.131 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:22.15007) ---- Analyzing Chunk 50 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2841 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 49.387   0.520  50.077 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:23.096671) ---- Analyzing Chunk 51 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2903 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 50.311   0.525  51.006 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:24.026932) ---- Analyzing Chunk 52 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2998 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 51.276   0.531  51.977 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:24.99765) ---- Analyzing Chunk 53 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2901 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 52.218   0.536  52.925 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:25.945443) ---- Analyzing Chunk 54 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3072 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 53.165   0.541  53.876 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:26.898716) ---- Analyzing Chunk 55 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2981 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 54.101   0.547  54.818 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:27.839016) ---- Analyzing Chunk 56 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3137 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 55.078   0.553  55.802 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:28.820837) ---- Analyzing Chunk 57 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3069 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 56.031   0.559  56.761 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:29.786902) ---- Analyzing Chunk 58 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2937 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 56.964   0.569  57.704 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:30.72379) ---- Analyzing Chunk 59 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3055 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 57.902   0.575  58.649 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:31.668431) ---- Analyzing Chunk 60 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2871 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 58.811   0.580  59.564 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:32.583916) ---- Analyzing Chunk 61 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2979 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 59.771   0.587  60.530 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:33.549139) ---- Analyzing Chunk 62 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2924 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 60.708   0.592  61.472 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:34.493667) ---- Analyzing Chunk 63 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3196 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 61.682   0.597  62.452 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:35.471631) ---- Analyzing Chunk 64 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3028 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 62.628   0.602  63.404 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:36.42344) ---- Analyzing Chunk 65 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2831 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 63.541   0.606  64.320 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:37.339145) ---- Analyzing Chunk 66 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3163 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 64.507   0.610  65.291 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:38.310764) ---- Analyzing Chunk 67 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3221 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 65.469   0.617  66.260 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:39.280386) ---- Analyzing Chunk 68 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2838 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 66.394   0.623  67.192 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:40.212016) ---- Analyzing Chunk 69 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2769 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 67.292   0.629  68.095 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:41.115721) ---- Analyzing Chunk 70 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3055 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 68.277   0.640  69.109 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:42.134704) ---- Analyzing Chunk 71 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2986 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 69.262   0.650  70.104 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:43.130907) ---- Analyzing Chunk 72 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3027 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 70.223   0.663  71.095 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:44.116902) ---- Analyzing Chunk 73 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3088 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 71.190   0.670  72.069 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:45.090224) ---- Analyzing Chunk 74 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3198 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 72.146   0.675  73.030 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:46.050782) ---- Analyzing Chunk 75 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2890 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 73.056   0.681  73.945 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:46.964768) ---- Analyzing Chunk 76 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3071 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 74.121   0.702  75.079 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:48.112973) ---- Analyzing Chunk 77 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3000 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 75.090   0.714  76.070 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:49.096461) ---- Analyzing Chunk 78 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2974 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 76.051   0.722  77.038 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:50.059445) ---- Analyzing Chunk 79 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3327 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 76.999   0.728  77.992 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:51.015646) ---- Analyzing Chunk 80 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2728 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 77.935   0.733  78.933 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:51.955705) ---- Analyzing Chunk 81 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3028 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 78.889   0.741  79.896 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:52.919691) ---- Analyzing Chunk 82 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3038 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 79.847   0.748  80.859 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:53.879427) ---- Analyzing Chunk 83 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2874 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 80.778   0.757  81.837 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:54.856341) ---- Analyzing Chunk 84 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2751 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 81.691   0.763  82.755 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:55.774718) ---- Analyzing Chunk 85 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3026 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 82.603   0.768  83.672 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:56.691608) ---- Analyzing Chunk 86 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3135 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 83.562   0.772  84.637 \n",
      "isVcfEnd  FALSE \n",
      "(2025-01-12 00:12:57.655823) ---- Analyzing Chunk 87 :  chrom InitialChunk ---- \n",
      "Reach the end of the vcf file\n",
      "1334 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 83.972   0.775  85.048 \n",
      "isVcfEnd  TRUE \n",
      "[1] \"Analysis done! The results have been saved to './chr21_10_4.bcf_results.txt'.\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: RhpcBLASctl\n",
      "Warning message:\n",
      "package ‘RhpcBLASctl’ was built under R version 4.3.3 \n",
      "Warning message:\n",
      "package ‘optparse’ was built under R version 4.3.3 \n",
      "Warning message:\n",
      "package ‘data.table’ was built under R version 4.3.1 \n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "Warning: skipping multiallelic variant\n",
      "IOPub message rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_msg_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_msg_rate_limit=1000.0 (msgs/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "export PATH=\"/opt/miniconda3/bin:$PATH\"\n",
    "conda run -n RSAIGE Rscript SAIGE/extdata/step2_SPAtests.R        \\\n",
    "        --vcfFile=../scaling/data/chr21_10_4.bcf \\\n",
    "        --vcfFileIndex=../scaling/data/chr21_10_4.bcf.csi \\\n",
    "        --vcfField=GT   \\\n",
    "        --SAIGEOutputFile=./chr21_10_4.bcf_results.txt \\\n",
    "        --chrom=1       \\\n",
    "        --minMAF=0 \\\n",
    "        --minMAC=20 \\\n",
    "        --GMMATmodelFile=./chr21_10_4.model.rda \\\n",
    "        --varianceRatioFile=./chr21_10_4.model.varianceRatio.txt  \\\n",
    "        --is_Firth_beta=TRUE    \\\n",
    "        --pCutoffforFirth=0.05 \\\n",
    "        --is_output_moreDetails=TRUE    \\\n",
    "        --LOCO=FALSE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f8bcf30-08c1-4635-948e-bd7c246eeb9d",
   "metadata": {},
   "source": [
    "Now we run [step 2](https://saigegit.github.io/SAIGE-doc/docs/single_step2.html) of the single-variant association test using the VCF Zarr data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8de12f7d-0c82-40ca-998d-2d58d0dccb5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R version 4.3.0 (2023-04-21)\n",
      "Platform: x86_64-apple-darwin13.4.0 (64-bit)\n",
      "Running under: macOS Big Sur ... 10.16\n",
      "\n",
      "Matrix products: default\n",
      "BLAS:   /opt/miniconda3/envs/RSAIGE/lib/libblis.4.0.0.dylib \n",
      "LAPACK: /opt/miniconda3/envs/RSAIGE/lib/liblapack.3.9.0.dylib\n",
      "\n",
      "locale:\n",
      "[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8\n",
      "\n",
      "time zone: UTC\n",
      "tzcode source: system (macOS)\n",
      "\n",
      "attached base packages:\n",
      "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
      "\n",
      "other attached packages:\n",
      "[1] data.table_1.15.2   optparse_1.7.5      RhpcBLASctl_0.23-42\n",
      "[4] SAIGE_1.3.6        \n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      "[1] compiler_4.3.0     Matrix_1.6-5       Rcpp_1.0.13        getopt_1.20.4     \n",
      "[5] grid_4.3.0         RcppParallel_5.1.9 lattice_0.22-6    \n",
      "$vcfFile\n",
      "[1] \"\"\n",
      "\n",
      "$vcfFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$vcfField\n",
      "[1] \"GT\"\n",
      "\n",
      "$vczFile\n",
      "[1] \"/Users/willtyler/Desktop/vcf-zarr-publication/scaling/data/chr21_10_4.zarr\"\n",
      "\n",
      "$savFile\n",
      "[1] \"\"\n",
      "\n",
      "$savFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$bgenFile\n",
      "[1] \"\"\n",
      "\n",
      "$bgenFileIndex\n",
      "[1] \"\"\n",
      "\n",
      "$sampleFile\n",
      "[1] \"\"\n",
      "\n",
      "$bedFile\n",
      "[1] \"\"\n",
      "\n",
      "$bimFile\n",
      "[1] \"\"\n",
      "\n",
      "$famFile\n",
      "[1] \"\"\n",
      "\n",
      "$AlleleOrder\n",
      "[1] \"alt-first\"\n",
      "\n",
      "$idstoIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$rangestoIncludeFile\n",
      "[1] \"\"\n",
      "\n",
      "$chrom\n",
      "[1] \"1\"\n",
      "\n",
      "$is_imputed_data\n",
      "[1] FALSE\n",
      "\n",
      "$minMAF\n",
      "[1] 0\n",
      "\n",
      "$minMAC\n",
      "[1] 20\n",
      "\n",
      "$minGroupMAC_in_BurdenTest\n",
      "[1] 5\n",
      "\n",
      "$minInfo\n",
      "[1] 0\n",
      "\n",
      "$maxMissing\n",
      "[1] 0.15\n",
      "\n",
      "$impute_method\n",
      "[1] \"best_guess\"\n",
      "\n",
      "$LOCO\n",
      "[1] FALSE\n",
      "\n",
      "$GMMATmodelFile\n",
      "[1] \"./chr21_10_4.model.rda\"\n",
      "\n",
      "$varianceRatioFile\n",
      "[1] \"./chr21_10_4.model.varianceRatio.txt\"\n",
      "\n",
      "$SAIGEOutputFile\n",
      "[1] \"./chr21_10_4.vcz_results.txt\"\n",
      "\n",
      "$markers_per_chunk\n",
      "[1] 10000\n",
      "\n",
      "$groups_per_chunk\n",
      "[1] 100\n",
      "\n",
      "$is_output_moreDetails\n",
      "[1] TRUE\n",
      "\n",
      "$is_overwrite_output\n",
      "[1] TRUE\n",
      "\n",
      "$maxMAF_in_groupTest\n",
      "[1] \"0.0001,0.001,0.01\"\n",
      "\n",
      "$maxMAC_in_groupTest\n",
      "[1] \"0\"\n",
      "\n",
      "$annotation_in_groupTest\n",
      "[1] \"lof,missense;lof,missense;lof;synonymous\"\n",
      "\n",
      "$groupFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMFile\n",
      "[1] \"\"\n",
      "\n",
      "$sparseGRMSampleIDFile\n",
      "[1] \"\"\n",
      "\n",
      "$relatednessCutoff\n",
      "[1] 0\n",
      "\n",
      "$sampleFile_male\n",
      "[1] \"\"\n",
      "\n",
      "$X_PARregion\n",
      "[1] \"\"\n",
      "\n",
      "$is_rewrite_XnonPAR_forMales\n",
      "[1] FALSE\n",
      "\n",
      "$MACCutoff_to_CollapseUltraRare\n",
      "[1] 10\n",
      "\n",
      "$cateVarRatioMinMACVecExclude\n",
      "[1] \"10,20.5\"\n",
      "\n",
      "$cateVarRatioMaxMACVecInclude\n",
      "[1] \"20.5\"\n",
      "\n",
      "$weights.beta\n",
      "[1] \"1,25\"\n",
      "\n",
      "$r.corr\n",
      "[1] 0\n",
      "\n",
      "$markers_per_chunk_in_groupTest\n",
      "[1] 100\n",
      "\n",
      "$condition\n",
      "[1] \"\"\n",
      "\n",
      "$SPAcutoff\n",
      "[1] 2\n",
      "\n",
      "$dosage_zerod_cutoff\n",
      "[1] 0.2\n",
      "\n",
      "$dosage_zerod_MAC_cutoff\n",
      "[1] 10\n",
      "\n",
      "$is_single_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_no_weight_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_output_markerList_in_groupTest\n",
      "[1] FALSE\n",
      "\n",
      "$is_Firth_beta\n",
      "[1] TRUE\n",
      "\n",
      "$pCutoffforFirth\n",
      "[1] 0.05\n",
      "\n",
      "$is_fastTest\n",
      "[1] FALSE\n",
      "\n",
      "$max_MAC_for_ER\n",
      "[1] 4\n",
      "\n",
      "$subSampleFile\n",
      "[1] \"\"\n",
      "\n",
      "$help\n",
      "[1] FALSE\n",
      "\n",
      "[1] \"opt$r.corr\"\n",
      "[1] 0\n",
      "dosage_zerod_cutoff  0.2 \n",
      "Any dosages <=  0.2  for genetic variants with MAC <=  10  are set to be 0.\n",
      "single-variant association test will be performed\n",
      "[1] \"Leave-one-chromosome-out is not applied\"\n",
      "[1] \"LOCO = FASLE and leave-one-chromosome-out is not applied\"\n",
      "variance Ratio null is  1.001513 \n",
      "dosageFile type is  vcz \n",
      "(2025-01-11 22:25:25.108538) ---- Analyzing Chunk 1 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3333 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  2.191   0.349   2.291 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:25.926401) ---- Analyzing Chunk 2 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3421 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  3.046   0.516   3.085 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:26.721993) ---- Analyzing Chunk 3 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3182 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  3.925   0.649   4.063 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:27.752188) ---- Analyzing Chunk 4 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3268 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  4.823   1.070   4.941 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:28.576485) ---- Analyzing Chunk 5 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3040 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  5.712   1.224   5.682 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:29.315375) ---- Analyzing Chunk 6 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3117 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  6.597   1.467   6.433 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:30.065598) ---- Analyzing Chunk 7 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3051 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  7.490   1.720   7.198 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:30.832474) ---- Analyzing Chunk 8 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3044 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  8.386   1.885   7.933 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:31.56611) ---- Analyzing Chunk 9 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3038 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      "  9.273   2.116   8.683 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:32.316233) ---- Analyzing Chunk 10 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3237 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 10.177   2.337   9.446 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:33.078637) ---- Analyzing Chunk 11 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3119 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 11.069   2.569  10.199 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:33.833146) ---- Analyzing Chunk 12 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3276 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 11.974   2.790  10.962 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:34.595083) ---- Analyzing Chunk 13 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3181 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 12.861   3.061  11.728 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:35.360382) ---- Analyzing Chunk 14 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3029 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 13.760   3.268  12.471 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:36.103774) ---- Analyzing Chunk 15 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2879 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 14.623   3.503  13.213 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:36.847552) ---- Analyzing Chunk 16 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3015 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 15.511   3.763  13.978 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:37.61088) ---- Analyzing Chunk 17 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3072 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 16.400   4.018  14.738 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:38.370524) ---- Analyzing Chunk 18 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3157 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 17.301   4.271  15.507 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:39.139912) ---- Analyzing Chunk 19 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3475 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 18.223   4.529  16.301 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:39.93477) ---- Analyzing Chunk 20 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3226 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 19.122   4.723  17.051 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:40.684151) ---- Analyzing Chunk 21 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3284 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 20.038   4.974  17.842 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:41.475476) ---- Analyzing Chunk 22 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3073 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 20.935   5.213  18.605 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:42.237512) ---- Analyzing Chunk 23 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2956 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 21.815   5.400  19.346 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:42.981388) ---- Analyzing Chunk 24 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3108 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 22.714   5.659  20.123 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:43.756576) ---- Analyzing Chunk 25 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3073 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 23.599   5.930  20.883 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:44.517727) ---- Analyzing Chunk 26 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2642 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 24.458   6.207  21.629 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:45.262151) ---- Analyzing Chunk 27 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2815 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 25.337   6.482  22.383 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:46.018292) ---- Analyzing Chunk 28 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2967 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 26.212   6.780  23.153 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:46.787377) ---- Analyzing Chunk 29 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2936 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 27.079   7.064  23.910 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:47.542384) ---- Analyzing Chunk 30 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2634 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 27.950   7.269  24.636 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:48.268957) ---- Analyzing Chunk 31 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3025 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 28.838   7.540  25.407 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:49.041896) ---- Analyzing Chunk 32 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3067 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 29.740   7.724  26.151 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:49.783736) ---- Analyzing Chunk 33 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3308 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 30.646   7.995  26.933 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:50.565282) ---- Analyzing Chunk 34 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3047 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 31.547   8.164  27.678 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:51.310253) ---- Analyzing Chunk 35 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3109 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 32.468   8.363  28.439 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:52.07282) ---- Analyzing Chunk 36 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2790 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 33.356   8.528  29.163 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:52.797049) ---- Analyzing Chunk 37 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3031 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 34.233   8.796  29.920 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:53.553409) ---- Analyzing Chunk 38 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3079 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 35.111   9.076  30.676 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:54.311533) ---- Analyzing Chunk 39 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3113 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 36.009   9.329  31.441 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:55.074457) ---- Analyzing Chunk 40 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3109 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 36.931   9.505  32.191 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:55.825308) ---- Analyzing Chunk 41 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3324 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 37.814   9.665  32.948 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:56.580774) ---- Analyzing Chunk 42 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3662 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 38.772   9.937  33.774 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:57.406713) ---- Analyzing Chunk 43 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2965 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 39.670  10.135  34.515 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:58.147596) ---- Analyzing Chunk 44 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2883 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 40.548  10.381  35.252 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:58.884798) ---- Analyzing Chunk 45 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3064 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 41.429  10.650  36.008 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:25:59.640937) ---- Analyzing Chunk 46 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3094 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 42.332  10.847  36.748 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:00.380692) ---- Analyzing Chunk 47 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3085 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 43.237  11.027  37.487 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:01.11991) ---- Analyzing Chunk 48 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2926 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 44.118  11.291  38.234 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:01.866999) ---- Analyzing Chunk 49 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2852 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 45.012  11.519  38.984 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:02.616918) ---- Analyzing Chunk 50 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2841 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 45.917  11.742  39.740 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:03.372551) ---- Analyzing Chunk 51 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2903 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 46.797  12.014  40.498 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:04.130783) ---- Analyzing Chunk 52 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2998 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 47.702  12.210  41.246 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:04.879574) ---- Analyzing Chunk 53 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2901 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 48.580  12.492  42.000 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:05.632071) ---- Analyzing Chunk 54 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3072 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 49.475  12.756  42.768 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:06.400667) ---- Analyzing Chunk 55 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2981 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 50.347  13.005  43.522 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:07.154866) ---- Analyzing Chunk 56 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3137 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 51.240  13.280  44.294 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:07.926243) ---- Analyzing Chunk 57 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3069 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 52.140  13.499  45.053 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:08.685315) ---- Analyzing Chunk 58 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2937 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 53.018  13.762  45.805 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:09.43721) ---- Analyzing Chunk 59 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3055 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 53.922  13.970  46.553 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:10.185828) ---- Analyzing Chunk 60 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2871 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 54.809  14.163  47.281 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:10.914632) ---- Analyzing Chunk 61 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2979 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 55.708  14.357  48.024 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:11.656828) ---- Analyzing Chunk 62 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2924 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 56.598  14.584  48.770 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:12.402488) ---- Analyzing Chunk 63 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3196 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 57.497  14.840  49.541 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:13.174157) ---- Analyzing Chunk 64 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3028 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 58.404  15.042  50.295 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:13.927648) ---- Analyzing Chunk 65 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2831 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 59.289  15.293  51.043 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:14.675687) ---- Analyzing Chunk 66 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3163 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 60.184  15.566  51.814 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:15.446293) ---- Analyzing Chunk 67 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3221 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 61.084  15.821  52.582 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:16.214965) ---- Analyzing Chunk 68 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2838 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 61.965  16.013  53.325 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:16.959655) ---- Analyzing Chunk 69 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2769 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 62.850  16.230  54.062 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:17.695881) ---- Analyzing Chunk 70 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3055 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 63.750  16.445  54.815 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:18.447615) ---- Analyzing Chunk 71 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2986 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 64.662  16.649  55.572 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:19.205217) ---- Analyzing Chunk 72 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3027 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 65.554  16.860  56.324 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:19.956261) ---- Analyzing Chunk 73 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3088 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 66.474  17.071  57.091 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:20.72353) ---- Analyzing Chunk 74 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3198 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 67.378  17.327  57.865 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:21.497041) ---- Analyzing Chunk 75 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2890 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 68.271  17.558  58.616 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:22.250002) ---- Analyzing Chunk 76 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3071 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 69.151  17.834  59.378 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:23.016162) ---- Analyzing Chunk 77 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3000 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 70.041  18.054  60.126 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:23.758243) ---- Analyzing Chunk 78 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2974 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 70.957  18.239  60.875 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:24.507663) ---- Analyzing Chunk 79 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3327 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 71.858  18.491  61.644 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:25.276347) ---- Analyzing Chunk 80 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2728 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 72.739  18.704  62.376 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:26.008921) ---- Analyzing Chunk 81 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3028 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 73.628  18.922  63.136 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:26.768561) ---- Analyzing Chunk 82 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3038 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 74.524  19.188  63.904 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:27.537033) ---- Analyzing Chunk 83 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2874 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 75.411  19.435  64.658 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:28.290506) ---- Analyzing Chunk 84 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "2751 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 76.283  19.674  65.391 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:29.024207) ---- Analyzing Chunk 85 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3026 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 77.172  19.848  66.131 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:29.763107) ---- Analyzing Chunk 86 :  chrom InitialChunk ---- \n",
      "Completed 10000/10000 markers in the chunk.\n",
      "3135 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 78.078  20.052  66.880 \n",
      "isVczEnd  FALSE \n",
      "(2025-01-11 22:26:30.517872) ---- Analyzing Chunk 87 :  chrom InitialChunk ---- \n",
      "1334 markers were tested.\n",
      "write to output\n",
      "   user  system elapsed \n",
      " 78.589  20.113  67.230 \n",
      "isVczEnd  TRUE \n",
      "[1] \"Analysis done! The results have been saved to './chr21_10_4.vcz_results.txt'.\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: RhpcBLASctl\n",
      "Warning message:\n",
      "package ‘RhpcBLASctl’ was built under R version 4.3.3 \n",
      "Warning message:\n",
      "package ‘optparse’ was built under R version 4.3.3 \n",
      "Warning message:\n",
      "package ‘data.table’ was built under R version 4.3.1 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "export PATH=\"/opt/miniconda3/bin:$PATH\"\n",
    "conda run -n RSAIGE Rscript SAIGE/extdata/step2_SPAtests.R        \\\n",
    "        --vczFile=/Users/willtyler/Desktop/vcf-zarr-publication/scaling/data/chr21_10_4.zarr \\\n",
    "        --vcfField=GT   \\\n",
    "        --SAIGEOutputFile=./chr21_10_4.vcz_results.txt \\\n",
    "        --chrom=1       \\\n",
    "        --minMAF=0 \\\n",
    "        --minMAC=20 \\\n",
    "        --GMMATmodelFile=./chr21_10_4.model.rda \\\n",
    "        --varianceRatioFile=./chr21_10_4.model.varianceRatio.txt  \\\n",
    "        --is_Firth_beta=TRUE    \\\n",
    "        --pCutoffforFirth=0.05 \\\n",
    "        --is_output_moreDetails=TRUE    \\\n",
    "        --LOCO=FALSE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd596404-b12c-4cb2-b509-4424b46f700c",
   "metadata": {},
   "source": [
    "### Compare results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fd76fb0f-b7b5-4dec-8f02-e926f351215d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "vcz_results = pd.read_csv('chr21_10_4.vcz_results.txt', sep='\\t')\n",
    "bcf_results = pd.read_csv('chr21_10_4.bcf_results.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "40af759b-7c30-4849-a3ce-28da261e1a8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcz_results.shape == bcf_results.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ab96228a-d5f2-44d5-846d-a31f4d3aa69a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all(vcz_results.columns == bcf_results.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "66847b07-88ae-4619-9b08-d66122ab99e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHR True\n",
      "POS True\n",
      "MarkerID False\n",
      "Allele1 False\n",
      "Allele2 False\n",
      "AC_Allele2 True\n",
      "AF_Allele2 True\n",
      "MissingRate True\n",
      "BETA True\n",
      "SE True\n",
      "Tstat True\n",
      "var True\n",
      "p.value True\n",
      "N True\n"
     ]
    }
   ],
   "source": [
    "for column in vcz_results.columns:\n",
    "    print(column, all(vcz_results[column] == bcf_results[column]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcbc358d-252c-4b68-93f7-40afc85128a2",
   "metadata": {},
   "source": [
    "### Cleanup\n",
    "\n",
    "A script, `cleanup.sh`, is added in the same folder as this notebook for convenience."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
